ractor.rb

```diff
@@ -1,34 +1,180 @@
+#     Client starts: #<Ractor:#3 test.rb:8 running>
+#     Client received from #<Ractor:#2 rac.rb:1 blocking>: ping
+#     Client sends to #<Ractor:#2 rac.rb:1 blocking>: pong
+#     Server received: pong
+#
+# It is said that Ractor receives messages via the <em>incoming port</em>, and sends them
+# to the <em>outgoing port</em>. Either one can be disabled with Ractor.close_incoming and
+# Ractor.close_outgoing respectively.
+#
+# == Shareable and unshareable objects
+#
+# When the object is sent to and from the ractor, it is important to understand whether the
+# object is shareable or not. Shareable objects are basically those which can be used by several
+# threads without compromising thread-safety; e.g. immutable ones. Ractor.shareable? allows to
+# check this, and Ractor.make_shareable tries to make object shareable if it is not.
+#
+#     Ractor.shareable?(1)            #=> true -- numbers and other immutable basic values are
+#     Ractor.shareable?('foo')        #=> false, unless the string is frozen due to # freeze_string_literals: true
+#     Ractor.shareable?('foo'.freeze) #=> true
+#
+#     str = 'foo'
+#     str.frozen?                 #=> false
+#     Ractor.make_shareable(str)
+#     str.frozen?                 #=> true
+#
+# When a shareable object is sent (via #send or Ractor.yield), no additional processing happens,
+# and it just becomes usable by both ractors. When an unshareable object is sent, it can be
+# either _copied_ or _moved_. The first is the default, and it makes the object's full copy by
+# deep cloning of its structure.
+#
+#     data = ['foo', 'bar']
+#     Ractor.new(data){
+#       data_in_ractor = Ractor.receive
+#       puts "In ractor: #{data_in_ractor.object_id}, #{data_in_ractor[0].object_id}"
+#     }
+#     r.send(data, move: true)
+#     sleep(0.1)
+#     puts "Outside: #{data.object_id}, #{data[0].object_id}"
+#
+# This will output:
+#
+#     In ractor: 100, 120
+#     Outside: 60, 80
+#
+# (Note that object id of both array and string inside array have changed inside
+# the ractor, showing it is different objects.)
+#
+# Deep cloning of the data may be slow, and sometimes impossible. Alternatively,
+# <tt>move: true</tt> may be used on sending. This will <em>move</em> the object to the
+# receiving ractor, making it inaccessible for a sending ractor.
+#
+#     data = ['foo', 'bar']
+#     r = Ractor.new{
+#       data_in_ractor = Ractor.receive
+#       puts "In ractor: #{data_in_ractor.object_id}, #{data_in_ractor[0].object_id}"
+#     }
+#     r.send(data, move: true)
+#     sleep(0.1)
+#     puts "Outside: moved? #{Ractor::MovedObject === data}"
+#     puts "Outside: #{data.inspect}"
+#
+# This will output:
+#
+#     In ractor: 100, 120
+#     Outside: moved? true
+#     test.rb:9:in `method_missing': can not send any methods to a moved object (Ractor::MovedError)
+#
+# Notice that even +inspect+ (and more basic methods like <tt>__id__</tt>) is inaccessible
+# on a moved object.
+#
```

Comment:
We need a section for other shareable objects: `Modules, Classes, Ractors`.

Reply:
Do we have a full list somewhere (in code?) The `doc/ractor.md` is a bit sparse on that matter.

Reply:
Code is not clear to me, but I believe that's it.
We should mention that thede are shareable but can not access instance variables from outside the main Ractor.

Reply:
Added. Also, I noticed I forgot to check and explain constant access rules, fixed it, too :)

Reply:
Funny that access to non-frozen constant value throws `NameError` and access to class instance vars throws `RuntimeError`, and not something Ractor-specific.